File: 2021/python/4.py
def find_first_winner(values, boards, matches):
    for value in values:
        for i in range(len(boards)):
            if value in boards[i]:
                matches[i][boards[i].index(value)] = 1

            if is_winning_board(matches[i]):
                unmarked = [boards[i][j] if not matches[i][j] else 0 for j in range(len(boards[i]))]
                return sum(unmarked) * value
    return -1


def find_last_winner(values, boards, matches):
    not_won_yet = list(range(len(boards)))
    last_winner = -1
    winning_value = -1
    for value in values:
        new_winners = []
        for i in not_won_yet:
            if value in boards[i]:
                matches[i][boards[i].index(value)] = 1

            if is_winning_board(matches[i]):
                new_winners.append(i)
                last_winner = i
                winning_value = value

        for winner in new_winners:
            not_won_yet.remove(winner)

    unmarked = [boards[last_winner][j] if not matches[last_winner][j] else 0 for j in range(len(boards[last_winner]))]
    return sum(unmarked) * winning_value


def print_board(board):
    for i in range(5):
        print(board[i*5:(i+1)*5])


def is_winning_board(matches):
    n_row, n_col = 5, 5
    for i in range(n_row):
        row = matches[i * n_col:(i + 1) * n_col]
        if sum(row) == 5:
            return True

    base_inds = list(range(0, len(matches), n_row))
    for i in range(n_col):
        col = [matches[ind + i] for ind in base_inds]
        if sum(col) == 5:
            return True

    return False


def main():
    boards = []
    matches = []
    with open("../inputs/4.txt", "r") as input:
        values = list(map(int, input.readline().split(",")))
        current_board = None
        for line in input:
            if line == "\n" and current_board is not None:
                boards.append(current_board)
                current_board = None
            else:
                new_row = list(map(int, line.strip("\n").split()))
                if current_board is None:
                    current_board = new_row
                else:
                    current_board.extend(new_row)
    if current_board is not None:
        boards.append(current_board)

    # boards = [boards[0]]

    matches = [[0 for i in range(len(boards[0]))] for j in range(len(boards))]

    print("PART ONE:", find_first_winner(values, boards, matches))

    # reset everything
    matches = [[0 for i in range(len(boards[0]))] for j in range(len(boards))]

    print("PART TWO:", find_last_winner(values, boards, matches))

main()

File: 2021/julia/4.jl
function find_first_winner(values, boards)
    # at the start there are no matches for each board
    n_boards, board_size = size(boards)
    matches = zeros(Int, (n_boards, board_size))

    # read out a value one at a time
    for i in eachindex(values)
        # check each board
        for j in 1:n_boards
            # if this value is on the board then mark it
            if values[i] in boards[j, :]
                matches[j, indexin(values[i], boards[j, :])[1]] = 1
            end;

            # check if this board just won
            if is_winning_board(matches[j, :])
                # get all of the unmarked values
                unmarked = boards[j, :] .* (1 .- matches[j, :])

                # sum them up and multiply by the winning value
                return sum(unmarked) * values[i]
            end;
        end;
    end;

    # return -1 if no one won
    return -1
end;


function find_last_winner(values, boards)
    # at the start there are no matches for each board
    n_boards, board_size = size(boards)
    matches = zeros(Int, (n_boards, board_size))

    # keep track of which boards haven't won yet
    playing = collect(1:n_boards)

    # Also who won last and what the value was
    last_winner = -1
    winning_value = -1

    # read out a value one at a time
    for i in eachindex(values)
        # track who won this round
        winners_this_round = []

        # check each board
        for j in eachindex(playing)
            # if this value is on the board then mark it
            if values[i] in boards[playing[j], :]
                matches[playing[j], indexin(values[i], boards[playing[j], :])[1]] = 1
            end;

            # check if this board just won
            if is_winning_board(matches[playing[j], :])
                winners_this_round = [winners_this_round;playing[j]]
                last_winner = playing[j]
                winning_value = values[i]
            end;
        end;

        for j in eachindex(winners_this_round)
            filter!(x -> x ≠ winners_this_round[j], playing)
        end;
    end;

    # count up the unmarked values in the last winner's board
    unmarked = boards[last_winner, :] .* (1 .- matches[last_winner, :])

    # multiply by winning value
    return sum(unmarked) * winning_value
end;


function is_winning_board(matches)
    """ Check if a board wins bingo"""
    n_row, n_col = 5, 5

    # check every row
    for i in 1:n_row
        row = matches[ (i - 1) * n_col + 1 : i * n_col]

        # if every value in the row matches then this board wins
        if sum(row) == n_col
            return true
        end;
    end;

    # check every column (more annoying than rows since not contiguous)
    base_inds = 1:n_row:n_col*n_row
    for i in 1:n_col
        col = [matches[ind + i - 1] for ind in base_inds]

        # if every value in the col matches then this board wins
        if sum(col) == n_row
            return true
        end;
    end;

    # otherwise this board hasn't won (yet!)
    return false
end;


function main()
    # start up an array of boards
    boards = Int[]
    values = Int[]
    open("../inputs/4.txt", "r") do input
        # get the first line of the file which contains the value that get read outs
        values = parse.(Int, split(chomp(readline(input)), ","))

        # start a new board
        current_board = nothing
        current_board = []
        for line in eachline(input)
            # if a new line occurs, save the current board and start a new one
            if chomp(line) == "" && current_board != []
                append!(boards, current_board)
                current_board = []
            # otherwise flatten out the board (combine rows)
            elseif chomp(line) != ""
                current_board = [current_board; parse.(Int, split(chomp(line)))]
            end;
        end;

        # don't miss the last board!
        append!(boards, current_board)
    end;

    boards = reshape(boards, (25, length(boards) ÷ 25)) |> transpose

    println("PART ONE:", find_first_winner(values, boards))
    println("PART TWO:", find_last_winner(values, boards))
end;

main()

